abi-versions/v0.2.1/README.md

```diff
@@ -0,0 +1,2166 @@
+[`proxy_on_http_call_response`], [`proxy_on_grpc_receive`]
+and/or [`proxy_on_queue_ready`] callbacks.
+
+Returned `status` value is:
+- `OK` on success.
+- `BAD_ARGUMENT` for unknown `context_id`.
+
+
+## Configuration
+
+### Callbacks exposed by the Wasm module
+
+#### `proxy_on_vm_start`
+
+* params:
+  - `i32 (uint32_t) unused`
+  - `i32 (size_t) vm_configuration_size`
+* returns:
+  - `i32 (bool) status`
+
+Called when the host starts the WebAssembly Virtual Machine.
+
+Its configuration (of `vm_configuration_size`) can be retrieved using
+[`proxy_get_buffer_bytes`] with `buffer_id` set to `VM_CONFIGURATION`.
+
+Plugin must return one of the following values:
+- `true` to indicate that the configuration was processed successfully.
+- `false` to indicate that the configuration processing failed, and that
+  this instance of WasmVM shouldn't be used.
+
+
+#### `proxy_on_configure`
+
+* params:
+  - `i32 (uint32_t) plugin_context_id`
+  - `i32 (size_t) plugin_configuration_size`
+* returns:
+  - `i32 (bool) status`
+
+Called when the host starts the Proxy-Wasm plugin.
+
+Its configuration (of `plugin_configuration_size`) can be retrieved
+using [`proxy_get_buffer_bytes`] with `buffer_id` set to
+`PLUGIN_CONFIGURATION`.
+
+Plugin must return one of the following values:
+- `true` to indicate that the configuration was processed successfully.
+- `false` to indicate that the configuration processing failed, and that
+  this instance of plugin shouldn't be used.
+
+
+## Logging
+
+### Functions exposed by the host
+
+#### `proxy_log`
+
+* params:
+  - `i32 (`[`proxy_log_level_t`]`) log_level`
+  - `i32 (const char *) message_data`
+  - `i32 (size_t) message_size`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Logs message (`message_data`, `message_size`) at the `log_level`.
+
+Returned `status` value is:
+- `OK` on success.
+- `BAD_ARGUMENT` for unknown `log_level`.
+- `INVALID_MEMORY_ACCESS` when `message_data` and/or `message_size`
+  point to invalid memory address.
+
+
+#### `wasi_snapshot_preview1.fd_write`
+
+* params:
+  - `i32 (`[`wasi_fd_id_t`]`) fd_id`
+  - `i32 (wasi_iovec_array_t) iovec`
+  - `i32 (size_t) iovec_size`
+  - `i32 (size_t *) return_written_bytes`
+* returns:
+  - `i32 (`[`wasi_errno_t`]`) errno`
+
+Logs message (`iovec`, `iovec_size`).
+
+When `fd_id` is `STDOUT`, then it's logged at the `INFO` level.
+
+When `fd_id` is `STDERR`, then it's logged at the `ERROR` level.
+
+Returned `errno` value is:
+- `SUCCESS` on success.
+- `BADF` for unknown or unsupported `fd_id`.
+- `FAULT` when `iovec`, `iovec_size` and/or `return_written_bytes`
+  point to invalid memory address.
+
+
+#### `proxy_get_log_level`
+
+* params:
+  - `i32 (`[`proxy_log_level_t`]` *) return_log_level`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Retrieves host's current log level (`return_log_level`).
+
+This can be used to avoid creating log entries that are going to be
+discarded by the host.
```

Comment:
note: the log level can be changed dynamically on the host.

Reply:
Added a note. Thanks!